### PR TITLE
Change the data loader method to load subset of data as generated by datamap from a previous run.

### PR DESCRIPTION
**Pull request by evanrex**
first pass done ~~but still need to test if working~~ and tested (is working)

1) fetch the .json file with training dynamics from the run that we want to use
2) get the data indices of the subset sample using the 'get_data_subset' function in training_dynamic_utils.py (probably needs some fiddling around to get it to work...havent tried it yet)
3) When fetching the data, we use the data indices to filter out points from the unshuffled training set
4) with the new, smaller dataset, training then remains the same

### CODE REVIEW

**Comment by evanrex** [comment]
~~TODO: test if actually works lmao~~ done

**Comment by timohromadka** [comment]
- In `utils/dataset_utils.py`
    - in `get_dataloaders()`, the `len(train_dataset)` is actually larger (50k) than the real training set (40k) that was used due to validation splitting. (e.g. 0.33 of hardtolearn data results in a dataset of size 16500 instead of 13200)
    - `tmp_test_subset = Subset(test_dataset, selected_indices)` works, even though the indices are almost completely out of range for the test set. Would you be able to just give it a look if the Subset functionality is working as expected, and consistently selecting the correct subset?
- In `utils/training_dynamic_utils.py`: would you be able to make a function which retrieves a data point, given an index, and displays it? This way we could manually verify if the 'hard to learn' data points actually do seem 'hard to learn'. Would be very cool to display these data points on the paper and see if it matches with human judgement on what is 'hard to learn'.

**Comment by evanrex** [comment]
>   * in `get_dataloaders()`, the `len(train_dataset)` is actually larger (50k) than the real training set (40k) that was used due to validation splitting. (e.g. 0.33 of hardtolearn data results in a dataset of size 16500 instead of 13200)

Are you sure? When I ran this yesterday I got 0.33 of variability to be the expected 13200. How did you get this result?

>  tmp_test_subset = Subset(test_dataset, selected_indices) works, even though the indices are almost completely out of range for the test set. Would you be able to just give it a look if the Subset functionality is working as expected, and consistently selecting the correct subset?

Ok I didn't check this so its likely the selected indices are buggy. I'll look into it

>  In utils/training_dynamic_utils.py: would you be able to make a function which retrieves a data point, given an index, and displays it? This way we could manually verify if the 'hard to learn' data points actually do seem 'hard to learn'. Would be very cool to display these data points on the paper and see if it matches with human judgement on what is 'hard to learn'.

Sure, will do

**Comment by evanrex** [comment]
> > * in `get_dataloaders()`, the `len(train_dataset)` is actually larger (50k) than the real training set (40k) that was used due to validation splitting. (e.g. 0.33 of hardtolearn data results in a dataset of size 16500 instead of 13200)
> 
> Are you sure? When I ran this yesterday I got 0.33 of variability to be the expected 13200. How did you get this result?


image

File: utils/dataset_utils.py [hunk 1]
```diff
@@ -69,7 +73,31 @@ def get_dataloaders(args):
...
+            selector_correctness=args.selector_correctness,  
+            p_forgetfulness=args.p_forgetfulness,
+            selector_forgetfulness=args.selector_forgetfulness  
+        )
+
+        # Create a Subset object for the selected indices
+        train_dataset = Subset(train_dataset, selected_indices)
     # Split the training dataset into train and validation
     train_size = int(0.8 * len(train_dataset))
     val_size = len(train_dataset) - train_size
```
Comment (timohromadka) [inline utils/dataset_utils.py]: Although `train_size` will now be 13200, it's not guaranteed to be the true 0.33 of the original 40k.

Also, the validation dataset should not be influenced by the subset selection. Perhaps the train/val split should be done deterministically for each run, and only then select the subset after the split.

File: utils/dataset_utils.py [hunk 2]
```diff
@@ -69,7 +73,31 @@ def get_dataloaders(args):
     #     # transform to spectrograms, or keep at waveform?
     else:
         raise ValueError("Unknown dataset")
-
+    
+    if args.prev_run_name_for_dynamics:
+        # get subset from dataset using previous run dynamics
+        gold_label_probabilities, confidence, variability, correctness, forgetfulness = get_training_dynamics_from_run_name(args.wandb_project_name, 'l46_datamaps', args.prev_run_name_for_dynamics)
+        selected_indices = get_data_subset(
+            list(range(len(train_dataset))),
```
Comment (timohromadka) [inline utils/dataset_utils.py:105]: Here, `len(train_dataset)` will be 50k as its working on the full dataset (train + val), thus `len(selected_indices)` will be 16500

File: utils/dataset_utils.py (hunk 1 above)
Comment (evanrex) [reply utils/dataset_utils.py]: Agreed, will do
